From f95ef5d3dad3e830c7794ac528cfd1f93a85a41e Mon Sep 17 00:00:00 2001
From: Pino Toscano <ptoscano@redhat.com>
Date: Fri, 31 Oct 2014 17:37:22 +0100
Subject: [PATCH 073/117] fish: fix dir completion on filesystems w/o
 dirent.d_type (RHBZ#1153844).

On filesystems whose dirent.d_type is DT_UNKNOWN or some unknown value,
manually check whether an entry is a directory, thus completing in the
proper way.

Partially cherry picked from commit a8b95a5535480c7b382fab82dcaf18eb67e5278a,
but adapted to manually resolving the path, since guestfs_is_dir uses
lstat.
---
 fish/destpaths.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/fish/destpaths.c b/fish/destpaths.c
index e79a90c..20ac56f 100644
--- a/fish/destpaths.c
+++ b/fish/destpaths.c
@@ -190,7 +190,13 @@ complete_dest_paths_generator (const char *text, int state)
                   else {
                     words = w;
                     words[nr_words].name = p;
-                    words[nr_words].is_dir = dirents->val[i].ftyp == 'd';
+                    if (dirents->val[i].ftyp == 'u'
+                        || dirents->val[i].ftyp == '?') {
+                      CLEANUP_FREE char *rp = guestfs_realpath (g, words[nr_words].name);
+                      if (rp)
+                        words[nr_words].is_dir = guestfs_is_dir (g, rp);
+                    } else
+                      words[nr_words].is_dir = dirents->val[i].ftyp == 'd';
                     nr_words++;
                   }
                 }
-- 
2.5.0

